src/librustc_codegen_utils/symbol_names/legacy.rs

```diff
@@ -112,7 +112,9 @@ fn get_symbol_hash<'tcx>(
 
         // also include any type parameters (for generic items)
         assert!(!substs.has_erasable_regions());
-        assert!(!substs.needs_subst());
+        if !tcx.sess.opts.debugging_opts.polymorphize {
+            assert!(!substs.needs_subst());
+        }
```

Comment:
Have you looked into `v0` mangling? You should be able to run most of the test suite with the default changed from `legacy` to `v0` in `src/librustc_session/options.rs`.